### PR TITLE
Use runtime storage in PoA -> Substrate module tests

### LINKED ISSUE #46: Test BridgeStorage trait
There are now 2 `trait Storage` implementors in PoA module: [`InMemoryStorage`](https://github.com/paritytech/parity-bridges-common/blob/b4877dbf87da88c843be57771c6a05d4cdd0426d/modules/ethereum/src/lib.rs#L765) and [`BridgeStorage`](https://github.com/paritytech/parity-bridges-common/blob/b4877dbf87da88c843be57771c6a05d4cdd0426d/modules/ethereum/src/lib.rs#L383). When the module was created, I've used in-memory storage to embed everything in parity-ethereum. #45 adds more logic to some storage methods && there are no actual tests for it.

So we either need to: (1) add tests for `BridgeStorage`, or (2) change all existing tests to use `BridgeStorage`.

### PR DESCRIPTION
**Pull request by svyatonik**
closes #46
part of #59

This PR removes `InMemoryStorage` + moves all Kovan references from runtime module to runtime level => now we can easily reconfgure it to work with any PoA chain.

### CODE REVIEW

File: modules/ethereum/src/finality.rs [hunk 1]
```diff
@@ -214,106 +214,110 @@ pub(crate) fn ancestry<'a, S: Storage>(
...
+					&Header::default(),
+					0,
+				),
+				Err(Error::NotValidator),
+			);
+		});
 	}
 
 	#[test]
 	fn prepares_votes() {
```
Comment (HCastano) [inline modules/ethereum/src/finality.rs:239]: I'd like to see all the stuff within `custom_test_ext()` be cleaned up with some functions, so that the test looks something like this:

```rust
	#[test]
	fn validators_finalize_block_when_threshold_is_met() {
		...

		let header1 = get_header(...);
		let finalized_blocks = get_finalized_blocks(header1, ...);
		assert_eq!(finalized_blocks, Ok(Vec::new()));

		let header2 = get_header(...);
		let finalized_blocks = get_finalized_blocks(header2, ...);
		assert_eq!(finalized_blocks, Ok(Vec::new()));

		let header3 = get_header(...);
		let finalized_blocks = get_finalized_blocks(header3, ...);
		assert_eq!(finalized_blocks, Ok(vec![(1, header1.hash, None)]);
	}

```

I found it hard to follow the meat of what the test was checking with all the setup involved.

File: modules/ethereum/src/import.rs [hunk 2]
```diff
@@ -166,214 +166,221 @@ pub fn header_import_requires_receipts<S: Storage>(
...
+			let mut storage = BridgeStorage::<TestRuntime>::new();
+			storage.finalize_headers(Some((100, Default::default())), None);
+			assert_eq!(
+				import_header(
+					&mut storage,
+					&test_aura_config(),
+					&test_validators_config(),
+					PRUNE_DEPTH,
+					None,
+					Default::default(),
```
Comment (HCastano) [inline modules/ethereum/src/import.rs:189]: Does this test work because the default block number would be 0? If so, it would be nice to make that more explicit

File: modules/ethereum/src/mock.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+/// Build default i-th block, using data from runtime storage.
+pub fn block_i(number: u64, validators: &[KeyPair]) -> Header {
+	custom_block_i(number, validators, |_| {})
+}
+
+/// Build custom i-th block, using data from runtime storage.
+pub fn custom_block_i(number: u64, validators: &[KeyPair], customize: impl FnOnce(&mut Header)) -> Header {
+	let validator_index: u8 = (number % (validators.len() as u64)) as _;
```
Comment (HCastano) [inline modules/ethereum/src/mock.rs:116]: Rust Q: What's `as _` meant to do?

File: modules/ethereum/src/mock.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	let mut header = Header {
+		number,
+		parent_hash: HeadersByNumber::get(number - 1).unwrap()[0].clone(),
+		gas_limit: 0x2000.into(),
+		author: validator(validator_index).address().to_fixed_bytes().into(),
```
Comment (HCastano) [inline modules/ethereum/src/mock.rs:121]: I've seen this a few times, could probably be put behind a function so we can just do `author: author(i)`

File: modules/ethereum/src/mock.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		seal: vec![vec![number as u8 + 42].into(), vec![].into()],
+		difficulty: number.into(),
+		..Default::default()
+	};
+	customize(&mut header);
+	signed_header(validators, header, number + 42)
```
Comment (HCastano) [inline modules/ethereum/src/mock.rs:127]: Does this 42 have anything to do with the seal, or something else?

File: modules/ethereum/src/mock.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+/// Return key pair of given test validator.
+pub fn validator(index: u8) -> KeyPair {
+	KeyPair::from_secret(Secret::from([index + 1; 32])).unwrap()
+}
+
+/// Return key pairs of all test validators.
+pub fn validators(count: u8) -> Vec<KeyPair> {
+	(0..count as usize).map(|index| validator(index as u8)).collect()
```
Comment (HCastano) [inline modules/ethereum/src/mock.rs:148]: Why's the `as usize` cast necessary here?

File: modules/ethereum/src/mock.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+use primitives::{Address, Header, H256, U256};
+use sp_runtime::{
+	testing::Header as SubstrateHeader,
+	traits::{BlakeTwo256, IdentityLookup},
+	Perbill,
+};
+
+pub type AccountId = u64;
+
+#[derive(Clone, Eq, PartialEq)]
```
Comment (HCastano) [inline modules/ethereum/src/mock.rs:32]:
```suggestion
#[derive(Clone, Eq, PartialEq, Debug)]
```

File: modules/ethereum/src/finality.rs [hunk 8]
```diff
@@ -214,106 +214,110 @@ pub(crate) fn ancestry<'a, S: Storage>(
...
-				0,
-			),
-			Err(Error::NotValidator),
-		);
+		custom_test_ext(genesis(), validators_addresses(5)).execute_with(|| {
+			assert_eq!(
+				finalize_blocks(
+					&BridgeStorage::<TestRuntime>::new(),
+					&Default::default(),
+					(&Default::default(), &[]),
```
Comment (HCastano) [inline modules/ethereum/src/finality.rs:227]: I'm confused as why this test returns `Error::NotValidator`, but I suspect is has to do with this. I think we should have a variable like this

```rust
// A valid author would be x, y, or z
let invalid_author: ... =  (&Default::default(), &[]);
```

to make it clear.

File: modules/ethereum/src/import.rs [hunk 9]
```diff
@@ -166,214 +166,221 @@ pub fn header_import_requires_receipts<S: Storage>(
...
+				import_header(
+					&mut storage,
+					&test_aura_config(),
+					&test_validators_config(),
+					PRUNE_DEPTH,
+					None,
+					block.clone(),
+					None,
+				)
+				.map(|_| ()),
```
Comment (HCastano) [inline modules/ethereum/src/import.rs:213]: Instead of doing this, could we not do `assert!(import_header(...).is_ok())`

File: modules/ethereum/src/import.rs [hunk 10]
```diff
@@ -166,214 +166,221 @@ pub fn header_import_requires_receipts<S: Storage>(
...
+				Err(Error::KnownHeader),
+			);
+		});
 	}
 
 	#[test]
 	fn import_header_works() {
```
Comment (HCastano) [inline modules/ethereum/src/import.rs:233]: I think this test is either doing too much, or the name needs to be updated.

If we're just checking that importing headers work, I don't think we need that `Multi` authority set thing that's going on. That could be part of a different test that checks validator set changes.

File: bin/node/runtime/src/kovan.rs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+/// Validators configuration for Kovan chain.
+pub fn kovan_validators_configuration() -> ValidatorsConfiguration {
+	ValidatorsConfiguration::Multi(vec![
+		(0, ValidatorsSource::List(kovan_genesis_validators())),
+		(
+			10960440,
+			ValidatorsSource::List(vec![
+				[
+					0x00, 0xD6, 0xCc, 0x1B, 0xA9, 0xcf, 0x89, 0xBD, 0x2e, 0x58, 0x00, 0x97, 0x41, 0xf4, 0xF7, 0x32,
```
Comment (HCastano) [inline bin/node/runtime/src/kovan.rs:43]: How are these generated?

File: modules/ethereum/src/verification.rs [hunk 12]
```diff
@@ -362,41 +367,88 @@ mod tests {
...
-		)
+		custom_test_ext(genesis(), validators_addresses(3)).execute_with(|| {
+			let validators = vec![validator(0), validator(1), validator(2)];
+			let mut storage = BridgeStorage::<TestRuntime>::new();
+			let block1 = block_i(1, &validators);
+			storage.insert_header(
+				storage
+					.import_context(None, &block1.parent_hash)
+					.unwrap()
+					.into_import_header(true, block1.hash(), block1, 0.into(), None, None),
```
Comment (HCastano) [inline modules/ethereum/src/verification.rs:391]: I think this can be done in a function. It would make the test flow in a similar way as before

File: modules/ethereum/src/verification.rs [hunk 13]
```diff
@@ -362,41 +367,88 @@ mod tests {
...
+				&test_aura_config(),
+				&validators_config,
+				&pool_configuration(),
+				&header,
+				receipts.as_ref(),
+			)
+		})
+	}
+
+	fn change_validators_set_at(number: u64, finalized_set: Vec<Address>, signalled_set: Option<Vec<Address>>) {
```
Comment (HCastano) [inline modules/ethereum/src/verification.rs:425]:
```suggestion
	fn change_validators_set_at(block_number: u64, finalized_set: Vec<Address>, signalled_set: Option<Vec<Address>>) {
```

**Comment by HCastano** [comment]
I realize that I may have made more comments to do with tests than the changes introduced in this PR. I wouldn't blame you if you don't want to fix them here (since they're not super relevant), but I think the tests should be cleaned up a bit at some point.

With that said, I actually do dig the changes, makes things a lot nicer to work with :)

File: modules/ethereum/src/mock.rs (hunk 6 above)
Comment (tomusdrw) [inline modules/ethereum/src/mock.rs:148]:
```suggestion
	(0..count).map(validator).collect()
```

File: modules/ethereum/src/mock.rs [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+	KeyPair::from_secret(Secret::from([index + 1; 32])).unwrap()
+}
+
+/// Return key pairs of all test validators.
+pub fn validators(count: u8) -> Vec<KeyPair> {
+	(0..count as usize).map(|index| validator(index as u8)).collect()
+}
+
+/// Return addresses of all test validators.
+pub fn validators_addresses(count: u8) -> Vec<Address> {
```
Comment (tomusdrw) [inline modules/ethereum/src/mock.rs:152]:
```suggestion
pub fn validator_addresses(count: u8) -> Vec<Address> {
```

double plural sounds okay for me (:slavic:), but afair it's not how you do it in english though :man_shrugging:

File: modules/ethereum/src/mock.rs [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+	(0..count as usize)
+		.map(|i| validator(i as u8).address().as_fixed_bytes().into())
```
Comment (tomusdrw) [inline modules/ethereum/src/mock.rs:154]: ditto

File: modules/ethereum/src/mock.rs (hunk 5 above)
Comment (svyatonik) [reply modules/ethereum/src/mock.rs:127]: IIRC it is some random offset to convert from block number to step number in tests

Comment (HCastano) [reply modules/ethereum/src/mock.rs:127]: We should make that something like `const BLOCK_TO_STEP_OFFSET = 42` to make it clear

File: modules/ethereum/src/import.rs (hunk 10 above)
Comment (HCastano) [reply modules/ethereum/src/import.rs:233]: Logged #111

File: modules/ethereum/src/finality.rs (hunk 1 above)
Comment (HCastano) [reply modules/ethereum/src/finality.rs:239]: Logged #111

File: modules/ethereum/src/import.rs (hunk 2 above)
Comment (HCastano) [reply modules/ethereum/src/import.rs:189]: Logged #111

File: modules/ethereum/src/import.rs (hunk 9 above)
Comment (HCastano) [reply modules/ethereum/src/import.rs:213]: Logged #112

**Comment by HCastano** [comment]
There are still some outstanding comments I have. One those are fixed/disputed I think this can be merged

File: modules/ethereum/src/mock.rs (hunk 5 above)
Comment (svyatonik) [reply modules/ethereum/src/mock.rs:127]: This is unrelated to this PR, because this fn has only been moved to that file. Can be addressed in #111

File: modules/ethereum/src/finality.rs (hunk 8 above)
Comment (svyatonik) [reply modules/ethereum/src/finality.rs:227]: Unrelated to this PR. Can be addressed in #111

**Comment by tomusdrw** [comment]
@HCastano can you take a look if all things were addressed?
